File: online-training/reservation-client/src/main/java/com/example/reservationclient/ReservationClientApplication.java
package com.example.reservationclient;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.rsocket.Payload;
import io.rsocket.RSocketFactory;
import io.rsocket.transport.netty.client.TcpClientTransport;
import io.rsocket.util.DefaultPayload;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.reactivestreams.Publisher;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.gateway.filter.ratelimit.RedisRateLimiter;
import org.springframework.cloud.gateway.route.RouteLocator;
import org.springframework.cloud.gateway.route.builder.RouteLocatorBuilder;
import org.springframework.cloud.netflix.hystrix.HystrixCommands;
import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpHeaders;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;

import java.io.IOException;

import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@SpringBootApplication
public class ReservationClientApplication {

	public static void main(String[] args) {
		SpringApplication.run(ReservationClientApplication.class, args);
	}

	@Bean
	RouterFunction<ServerResponse> routes(ReservationClient client) {
		return route(GET("/reservations/names"), serverRequest -> {

			Flux<String> names = client
				.getAllReservations()
				.map(Reservation::getName);

			Publisher<String> cb = HystrixCommands
				.from(names)
				.eager()
				.commandName("names")
				.fallback(Flux.just("EEK!"))
				.build();


		/*	// hedging
			WebClient wc = null; ///
			DiscoveryClient dc = null; ///
			List<ServiceInstance> instances = dc.getInstances("foo-service");
			if (instances.size() >= 3) {
				List<ServiceInstance> serviceInstances = instances.subList(0, 3);

				Flux<Reservation> callThreeServicesAtTheSameTime = Flux
					.fromStream(serviceInstances.stream())
					.map(si -> si.getHost() + ':' + si.getPort())
					.flatMap(uri -> wc.get().uri(uri).retrieve().bodyToFlux(Reservation.class));

				Flux<Reservation> first = Flux.first(callThreeServicesAtTheSameTime);

			}
*/
			return ServerResponse.ok().body(cb, String.class);
		});
	}

	@Bean
	WebClient webClient(WebClient.Builder builder) {
		return builder.build();
	}

	@Bean
	MapReactiveUserDetailsService authentication() {
		UserDetails jlong = User.withDefaultPasswordEncoder().username("jlong").password("pw").roles("USER").build();
		UserDetails rwinch = User.withDefaultPasswordEncoder().username("rwinch").password("pw").roles("USER", "ADMIN").build();
		return new MapReactiveUserDetailsService(jlong, rwinch);
	}

	@Bean
	SecurityWebFilterChain authorization(ServerHttpSecurity http) {
		http
			.authorizeExchange()
			.pathMatchers("/proxy").authenticated()
			.anyExchange().permitAll();
		http.httpBasic();
		http.csrf().disable();
		return http.build();
	}

	@Bean
	RedisRateLimiter redisRateLimiter() {
		return new RedisRateLimiter(5, 7);
	}

	@Bean
	RouteLocator gateway(RouteLocatorBuilder routeLocatorBuilder) {
		return routeLocatorBuilder
			.routes()
			.route(rs -> rs
				.host("*.foo.bar").and().path("/proxy")
				.filters(fSpec -> fSpec
					.setPath("/reservations")
					.addResponseHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, "*")
					.requestRateLimiter(rl -> rl.setRateLimiter(redisRateLimiter()))
				)
				.uri("http://localhost:8080")
			)
			.build();
	}
}
/*
@Component
class ReservationClient {

	private final WebClient webClient;

	ReservationClient(WebClient webClient) {
		this.webClient = webClient;
	}

	Flux<Reservation> getAllReservations() {
		return this.webClient
			.get()
			.uri("http://localhost:8080/reservations")
			.retrieve()
			.bodyToFlux(Reservation.class);
	}
}*/

@Component
class ReservationClient {

	private final TcpClientTransport tcpClientTransport = TcpClientTransport.create(7000);

	ReservationClient(ObjectMapper objectMapper) {
		this.objectMapper = objectMapper;
	}

	Flux<Reservation> getAllReservations() {
		return RSocketFactory
			.connect()
			.transport(this.tcpClientTransport)
			.start()
			.flatMapMany(rs ->
				rs.requestStream(DefaultPayload.create(new byte[0]))
					.map(Payload::getDataUtf8)
					.map(this::to)
			);
	}

	private Reservation to(String json) {
		try {
			return this.objectMapper.readValue(json, Reservation.class);
		}
		catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	private final ObjectMapper objectMapper;
}

@Data
@AllArgsConstructor
@NoArgsConstructor
class Reservation {
	private Integer id;
	private String name;
}

File: finchley/reservation-client-kotlin/src/main/kotlin/com/example/reservationclient/ReservationClientApplication.kt
package com.example.reservationclient

import org.reactivestreams.Publisher
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.cloud.client.loadbalancer.reactive.LoadBalancerExchangeFilterFunction
import org.springframework.cloud.gateway.discovery.DiscoveryClientRouteDefinitionLocator
import org.springframework.cloud.gateway.filter.factory.RequestRateLimiterGatewayFilterFactory
import org.springframework.cloud.gateway.filter.ratelimit.RedisRateLimiter
import org.springframework.cloud.gateway.route.builder.RouteLocatorBuilder
import org.springframework.cloud.gateway.route.builder.filters
import org.springframework.cloud.gateway.route.builder.routes
import org.springframework.cloud.netflix.hystrix.HystrixCommands
import org.springframework.cloud.stream.annotation.EnableBinding
import org.springframework.cloud.stream.messaging.Source
import org.springframework.context.support.beans
import org.springframework.messaging.support.MessageBuilder
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.userdetails.MapReactiveUserDetailsService
import org.springframework.security.core.userdetails.User
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToFlux
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.bodyToMono
import org.springframework.web.reactive.function.server.router
import reactor.core.publisher.Flux

@SpringBootApplication
@EnableBinding(Source::class)
class ReservationClientApplication

fun main(args: Array<String>) {
	SpringApplicationBuilder()
			.sources(ReservationClientApplication::class.java)
			.initializers(beans {
				bean {
					DiscoveryClientRouteDefinitionLocator(ref())
				}
				bean {
					MapReactiveUserDetailsService(
							User.withDefaultPasswordEncoder()
									.username("user")
									.roles("USER")
									.password("pw")
									.build())
				}
				bean {
					//@formatter:off
					val http = ref <ServerHttpSecurity>()
					http
							.csrf().disable()
							.httpBasic()
							.and()
							.authorizeExchange()
								.pathMatchers("/proxy").authenticated()
								.anyExchange().permitAll()
							.and()
							.build()
					//@formatter:on
				}
				bean {
					val builder = ref<RouteLocatorBuilder>()
					builder
							.routes {
								route {
									val rl = ref<RequestRateLimiterGatewayFilterFactory>()
											.apply(RedisRateLimiter.args(2, 4))
									path("/proxy")
									filters {
										filter(rl)
									}
									uri("lb://reservation-service/reservations")
								}
							}
				}
				bean {
					WebClient.builder().filter(ref<LoadBalancerExchangeFilterFunction>()).build()
				}
				bean {
					router {
						val writesChannel = ref<Source>()
						val client = ref<WebClient>()

						GET("/reservations/names") {
							val names: Publisher<String> = client
									.get()
									.uri("http://reservation-service/reservations")
									.retrieve()
									.bodyToFlux<Reservation>()
									.map { it.reservationName }

							val cb = HystrixCommands
									.from(names)
									.commandName("reservation-names")
									.fallback(Flux.just("EEEK!!"))
									.eager()
									.build()

							ServerResponse.ok().body(cb)
						}
						POST("/reservations") { req ->
							req.bodyToMono<Reservation>()
									.map { it.reservationName }
									.map { MessageBuilder.withPayload(it!!).build() }
									.map { writesChannel.output().send(it) }
									.flatMap { ServerResponse.ok().build() }
						}
					}
				}
			})
			.run(*args)
}

data class Reservation(var id: String? = null, var reservationName: String? = null)